src/cargo/core/compiler/mod.rs

```diff
@@ -726,6 +726,16 @@ fn prepare_rustc(
     if unit.target.is_test() || unit.target.is_bench() {
         let tmp = cx.files().layout(unit.kind).prepare_tmp()?;
         base.env("CARGO_TARGET_TMPDIR", tmp.display().to_string());
+
+        let cargo_manifest_dir = unit.pkg.root();
+        let cargo_workspace_dir = cx.bcx.ws.root();
+        let cargo_workspace_dir_rel_path =
+            pathdiff::diff_paths(cargo_workspace_dir, cargo_manifest_dir)
+                .expect("both paths are absolute");
+        base.env(
+            "CARGO_WORKSPACE_DIR",
+            cargo_workspace_dir_rel_path.display().to_string(),
+        );
```

Comment:
I read this <https://github.com/rust-lang/cargo/issues/3946#issuecomment-1402464796> and it suggests puttting this new environment variable behind nightly. So, maybe this?

```suggestion
        if cx.bcx.config.nightly_features_allowed {
            base.env(
                "CARGO_WORKSPACE_DIR",
                cargo_workspace_dir_rel_path.display().to_string(),
            );
        }
```

Reply:
This was on me.  I had forgotten about that and had suggested we start this way just for starting the conversation.